Backport ovs related additions from crowbar-core

rhafer opened the pull request:
Currently this also contains the commit from https://github.com/crowbar/barclamp-network/pull/328, but I'll rebase once that is merged. Or we can handle all of it here and close the other PR.

rhafer committed:
Move package list to default attributes

Cherry-picked from 18d9e362c4cbed9a06e69f4f89f9ba7ee95c5270 (crowbar-core)

rhafer committed:
Add ability to create/modify openvswitch bridges

Cherry-picked from 7dd36280f9e93133feeac9e41275a69b0dd9782c (crowbar-core)

rhafer committed:
Fixed HoundCI complaints

Cherry-picked from 57de7e03391bbc197e323ce103f0c5fb69dc0cc0 (crowbar-core)

rhafer committed:
Replug vlan in OVS after enslaving parent device

This is a workaround for bnc#945219 to avoid network connectivity issue.

Cherry-picked from 486c4cff7481f171febc7ae373c33a052152fa80 (crowbar-core)

rhafer committed:
Try to create and run the ovs resource only once

Cherry-picked from 8f3484bc967a283d957c1c5cf9839256290e55a5 (crowbar-core)

rhafer committed:
Only run replugging workaround when really needed (bsc#945219)

Avoid doing it on every chef-client run

Cherry-picked from 5f7af45e27c769697bab081b08a6b9c2b5f933a8 (crowbar-core)

rhafer committed:
Enable wicked-nanny when ovs bridges are managed by wicked

As documented on https://en.opensuse.org/Portal:Wicked/OpenvSwitch. There is
not need to restart wicked immediately as the ovs bridges on the running
enviroment where already correctly setup by crowbar.

Cherry-picked from 88a243d797e6e3edb3cc86fc696fc69ce5c63d12 (crowbar-core)

rhafer committed:
Move SLE11 ovs workaround init scripts to network barclamp

This is the network barclamp's part of moving the SLE11 workaround init scripts
to configure the IP on ovs bridges during boot (formerly know as
ovs-usurp-config-br-*) from the neutron to the network barclamp. As the network
barclamp now handles the creation the bridges and create ifcfg files for them
the new init scripts could also be kept very simple (they just call ifup).

rhafer commented:
@vuntz, @dirkmueller, @jdsn   as I am still not able to add Labels here could you please add the wip and/or donotmerge labels here?

rhafer commented on chef/cookbooks/network/recipes/default.rb:
```diff
@@ -453,6 +500,22 @@ def kill_nic(nic)
       end
     end
-
+    if node[:platform] == "suse" && node[:platform_version].to_f < 12.0
```
@vuntz I am specifically interested in feedback on this part here. Since it will significantly change how we handle SLE11 nodes. Are you ok with this. Or would you prefer the SLE11 code path unchanged. I've the slight feeling that the code in neutron barclamp will get really messy if we try to keep current SLE11 approach  (handling everything in the neutron cookbooks) working as it currently is.

vuntz replied:
I'm fine with changing this, as long as we fully test an upgrade from before the patch to after the patch.

vuntz commented:
> could you please add the wip and/or donotmerge labels here?

interestingly, the labels are missing here... Not sure why.

jdsn commented:
@rhafer I added the labels
I even found out that the labels are very inconsistent in all the repos, I guess I'll write a small script to sync them.

rhafer committed:
Set correct openvswitch service name for SLE12 nodes (bsc#935912)

https://bugzilla.suse.com/show_bug.cgi?id=935912

Cherry-picked from 77c120e5e7008d4fa2e98c725a7c814919b53f0d (crowbar-core)

rhafer committed:
Introduce new attribute to request openvswitch installation (bsc#950798)

Some nodes, e.g. compute nodes when using ovs+gre/vxlan, need openvswitch
running without requiring any bridges been created by the barclamp. This
commit decouples the installion of the packages and startup of the service
from the "add_ovs_bridge" attribute. And adds the new "needs_openvswitch"
attribute. This attribute in not present in the proposal as it's specific
to single nodes and not for the whole deployment of a network proposal.

https://bugzilla.suse.com/show_bug.cgi?id=950798

Cherry-picked from 557e1bb44a6e42543fd8842afebcf85bba8cd4f6 (crowbar-core)

rhafer committed:
Disable old openvswitch-switch service non SLE11 nodes

We're using the systemd unit file there now. Which has a different
name.

Cherry-picked from 4e7f9260c66d67d05c14e086f1f8447858238c32 (crowbar-core)

rhafer commented:
Update this with the latest fixes from crowbar-core. Still need to update the commit references once https://github.com/crowbar/crowbar-core/pull/105 is merged.

rhafer committed:
Fix ifcfg template for openvswitch device

According to the SLES12-SP1 documentation STARTMODE can be "auto" for
OvsBridges as well.
We need the OVS_BRIDGE='yes' and OVS_BRIDGE_PORT_DEVICE_* settings as wicked
will remove the bridge during shutdown and recreate them at boottime. (Which
went unnoticed until recently because of a bug in the openvswitch service
dependencies)

Cherry-picked from 8f27eb66e9974b3674eb3de764e0b54106f59c05 (crowbar-core)

rhafer commented:
I added the wip label back. Depending on what wicked  (and ovs packaging) changes are released for SLE12 we might need to backport the recent fixes from master as well. Need to check that first before we can merge anything here.

rhafer committed:
Don't create ovs-system related pieces anymore

They are no longer needed with the latest wicked (>= 0.6.28)

Cherry-picked from 5f55f228a7e38f69c71a954624ce1952ebfb0637 (crowbar-core)

rhafer committed:
Only use bootproto "none" for bond slaves

Should be "static" (with no IP Address) for OVS slaves. "none" causes reboot
problems at least on SLE11 nodes, but this change should be save for SLE12 as
well.

Cherry-picked from 64f8c1f99b840c9d5b3aa040987e3b41eb5f5eba (crowbar-core)

rhafer commented:
Rebased and imported the latest adjustments from crowbar-core. This is finally ready for review.

rhafer commented:
Note: it should be safe to merge this independently of: https://github.com/crowbar/barclamp-neutron/pull/260

mjura commented:
+1

aplanas commented:
+1

vuntz commented on chef/cookbooks/network/recipes/default.rb:
```diff
@@ -453,6 +511,47 @@ def kill_nic(nic)
+      service "ovs-ifup-#{nic.name}" do
+        # Don't start it here. It only needs to be executed during boot.
+        action [:nothing]
+        subscribes :enable, resources("template[/etc/init.d/ovs-ifup-#{nic.name}]")
```
What about implementing https://github.com/crowbar/barclamp-neutron/pull/262/files here?

rhafer replied:
Hm, right. Forgot about that one.

vuntz commented:
Except from the small comment, this looks good to me.

rhafer committed:
Do not rely on notification to enable ovs-usurp-config scripts

With a notification, if chef-client crashes before the end, the script
will never be enabled. Just always enable it, there's no reason why we
would want to not do this in the case where we have the script.

rhafer commented:
> Except from the small comment, this looks good to me.

@vuntz addressed.

vuntz commented:
+1
